### PR TITLE
Trying to get it working on Mobile.  Made many bug fixes along the way.

### PR DESCRIPTION
**Pull request by sharpninja**
## Bug Fixes 

* Changed `Dictionary` to `ConcurrentDictionary` wherever possible.
* Changed indexer accesses to `TryGetValue` everywhere
* Copied collections to temporary variables for iterating to prevent concurrency issues.
* Added `Microsoft.Extensions,DependencyInjection`
* Added `Microsoft.Extensions.Logging`
* __Added handling of  `InvalidSession` operation__

## Mobile

* `AcrylicBrush` is not supported on Mobile
* There are more theme resources being used that are also missing on Mobile.

### CODE REVIEW

File: src/Quarrel.ViewModels/Models/Bindables/BindableChannel.cs [hunk 1]
```diff
@@ -437,38 +438,44 @@ public bool IsTyping
...
+            List<string> names = new List<string>();
+            var keys = Typers.Keys.ToList();
+
+            foreach (var id in keys)
             {
-                List<string> names = new List<string>();
-                foreach (var id in Typers.Keys)
+                var typer = _CurrentUsersService.Users.TryGetValue(id, out var user) ? user.DisplayName : null;
+
+                if (typer != null)
```
Comment (matthew4850) [inline src/Quarrel.ViewModels/Models/Bindables/BindableChannel.cs]: Is there a reason why you choose to use a temporary variable instead of doing this?
if(TryGetvalue(id, out var user){
    names.add(user.DisplayName);
}

File: src/Quarrel.ViewModels/Extensions/System.Collections.Generic/HashedCollection.cs [hunk 2]
```diff
@@ -125,7 +125,7 @@ public bool ContainsKey(TKey key)
 
         public bool ContainsValue(TValue value)
         {
-            return _dict.ContainsValue(value);
+            return _dict.Values.Any(v => v.Equals(value));
         }
```
Comment (matthew4850) [inline src/Quarrel.ViewModels/Extensions/System.Collections.Generic/HashedCollection.cs]: What is the purpose of this change?

File: src/Quarrel/Quarrel.csproj [hunk 3]
```diff
@@ -18,11 +18,17 @@
...
-    <AppInstallerUpdateFrequency>1</AppInstallerUpdateFrequency>
+    <AppxBundlePlatforms>arm</AppxBundlePlatforms>
+    <AppInstallerUpdateFrequency>0</AppInstallerUpdateFrequency>
     <AppInstallerCheckForUpdateFrequency>OnApplicationRun</AppInstallerCheckForUpdateFrequency>
+    <AppxPackageSigningEnabled>True</AppxPackageSigningEnabled>
+    <PackageCertificateKeyFile>C:\GitHub\Quarrel\src\Quarrel\Quarrel_TemporaryKey.pfx</PackageCertificateKeyFile>
+    <PackageCertificateThumbprint>1E2AFBDDC11EE62EAB18FEABE0C936C5B666C533</PackageCertificateThumbprint>
+    <AppxPackageSigningTimestampDigestAlgorithm>SHA256</AppxPackageSigningTimestampDigestAlgorithm>
+    <GenerateTestArtifacts>True</GenerateTestArtifacts>
+    <HoursBetweenUpdateChecks>0</HoursBetweenUpdateChecks>
```
Comment (matthew4850) [inline src/Quarrel/Quarrel.csproj]: You left temporary keys in here, they should be removed.
Is there any reason why AppxBundlePlatforms was changed?
Why was AppxAutoIncrementPackageRevision enabled?

File: src/Quarrel/Quarrel.csproj [hunk 4]
```diff
@@ -482,11 +488,21 @@
     <Content Include="Assets\IconAssets\WideTile.scale-400.png" />
     <Content Include="Controls\Markdown\About.txt" />
     <Content Include="Controls\Markdown\oltheme.txt" />
+    <AdditionalFiles Include="appsettings.json" />
     <None Include="Package.StoreAssociation.xml" />
     <Content Include="Properties\Default.rd.xml" />
-    <Content Include="SodiumC.dll" />
+    <None Include="Quarrel_TemporaryKey.pfx" />
+  </ItemGroup>
```
Comment (matthew4850) [inline src/Quarrel/Quarrel.csproj]: Temporary key again.

File: src/Quarrel/App.xaml.cs [hunk 5]
```diff
@@ -42,15 +49,64 @@ namespace Quarrel
...
+            ServiceProvider = services.BuildServiceProvider();
+
+            Configuration = ServiceProvider.GetService<IConfiguration>();
+
+            var folder = ApplicationData.Current.LocalFolder;
+            string fullPath = $"{folder.Path}\\Logs\\App.log";
+
+            ServiceProvider.GetService<ILoggerFactory>().AddFile(fullPath, LogLevel.Debug);
+            ServiceProvider.GetService<ILoggerFactory>().AddDebug((s, l) => true);
+
```
Comment (matthew4850) [inline src/Quarrel/App.xaml.cs]: Why is the static constructor needed?

File: src/Quarrel/App.xaml.cs [hunk 6]
```diff
@@ -138,32 +198,68 @@ public void SetupTitleBar()
...
+                    try
+                    {
+                        statusBar.BackgroundOpacity = 1;
+                        statusBar.BackgroundColor = ((SolidColorBrush)Current.Resources["AcrylicCommandBarBackground"]).Color;
+                        statusBar.ForegroundColor = ((SolidColorBrush)Current.Resources["MessageForeground"]).Color;
+                    }
+                    catch (Exception ex) {
+                        Logger.LogError(new EventId(), ex, "Error caught accessing resources. (Group 1)");
+                        _lightOnly = true;
+                    }
```
Comment (matthew4850) [inline src/Quarrel/App.xaml.cs]: Won't this disable changing theme on mobile as this will always fail on mobile due to lack of acrylic support?

File: src/Quarrel/App.xaml [hunk 7]
```diff
@@ -5,74 +5,118 @@
...
                                   FallbackColor="#A0FFFFFF"/>
 
+                    <SolidColorBrush x:Key="LightBG" Color="#f6f7f9"/>
+                    <SolidColorBrush x:Key="MidBG" Color="#e3e8ed"/>
+                    <SolidColorBrush x:Key="MidBG_hover" Color="#2e3036"/>
+                    <SolidColorBrush x:Key="MidBG_focus" Color="#303238"/>
+                    <SolidColorBrush x:Key="DarkBG" Color="#e1e3e5"/>
+                    <SolidColorBrush x:Key="LinkColor" Color="#0096cf"/>
+                    <SolidColorBrush x:Key="InvertedBG" Color="#161b2b"/>
+                </Windows10FallCreatorsUpdate:ResourceDictionary>-->
```
Comment (matthew4850) [inline src/Quarrel/App.xaml]: Quarrel should maintain acrylic support on light mode, perhaps there should be four themes, light acrylic, dark acrylic, light and dark.

File: src/Quarrel.ViewModels/Services/Gateway/GatewayService.cs [hunk 8]
```diff
@@ -78,25 +84,32 @@ public async void InitializeGateway([NotNull] string accessToken)
...
+        }
 
-                await Gateway.SubscribeToGuild(idList.ToArray());
-            });
+        public async Task<bool> ConnectWithRetryAsync(int retriesLeft)
+        {
+            if (await Gateway.ConnectAsync()) return true;
+            else if (retriesLeft > 0) { return await ConnectWithRetryAsync(--retriesLeft); }
+            else return false;
         }
```
Comment (matthew4850) [inline src/Quarrel.ViewModels/Services/Gateway/GatewayService.cs]: I think a loop would look clearer here, why the use of recursion?

File: src/Quarrel.ViewModels/MainViewModel.cs [hunk 9]
```diff
@@ -287,32 +296,35 @@ private void RegisterMessages()
...
                 {
-                    var bChannel = GuildsService.CurrentChannels[m.TypingStart.ChannelId];
-                    if (!bChannel.Typers.ContainsKey(m.TypingStart.UserId))
+                    if (GuildsService.CurrentChannels.TryGetValue(m.TypingStart.ChannelId, out var bChannel))
                     {
-                        Timer timer = new Timer ((s) =>
+                        if (bChannel.Typers.TryRemove(m.TypingStart.UserId, out var oldTimer))
+                        {
+                            oldTimer.Dispose();
+                        }
```
Comment (matthew4850) [inline src/Quarrel.ViewModels/MainViewModel.cs]: Why dispose the old timer instead of updating it?

File: src/Quarrel/App.xaml.cs [hunk 10]
```diff
@@ -138,32 +198,68 @@ public void SetupTitleBar()
...
-                case Theme.Light:
-                    Application.Current.RequestedTheme = ApplicationTheme.Light;
-                    break;
+                Application.Current.RequestedTheme = ApplicationTheme.Dark;
+                var _ = Current.Resources["AcrylicCommandBarBackground"];
             }
+            catch
+            {
+                Application.Current.RequestedTheme = ApplicationTheme.Light;
+                _lightOnly = true;
```
[the remark this answers is not in the document]
Comment (sharpninja) [reply src/Quarrel/App.xaml.cs]: Light here doesn't include Acrylic (I removed it from the theme in App.xaml), so only Mobile users will default to the "light" them.  The switch block below seems to always fall through to default, even on 20H1.

File: src/Quarrel.ViewModels/Models/Bindables/BindableChannel.cs (hunk 1 above)
Comment (sharpninja) [reply src/Quarrel.ViewModels/Models/Bindables/BindableChannel.cs]: I was debugging and needed the local variable.  You can change it.

File: src/Quarrel.ViewModels/Extensions/System.Collections.Generic/HashedCollection.cs (hunk 2 above)
Comment (sharpninja) [reply src/Quarrel.ViewModels/Extensions/System.Collections.Generic/HashedCollection.cs]: ConcurrentDictionary<,> doesn't have ContainsValue.  The code is functionally equivalent as both must iterate and compare each Value until a match is found.

File: src/Quarrel/Quarrel.csproj (hunk 3 above)
Comment (sharpninja) [reply src/Quarrel/Quarrel.csproj]: To test o mobile I have to generate an appx for each build and manually deploy it via the device portal.  These are settings that can be broken out to another file and only updated locally.

File: src/Quarrel/Quarrel.csproj (hunk 4 above)
Comment (sharpninja) [reply src/Quarrel/Quarrel.csproj]: Yeah, oversight.  You can remove that line.

File: src/Quarrel/App.xaml.cs (hunk 5 above)
Comment (sharpninja) [reply src/Quarrel/App.xaml.cs]: By convention, Microsoft's IOC is initialized at the static context before the App is created.

File: src/Quarrel/App.xaml.cs (hunk 6 above)
Comment (sharpninja) [reply src/Quarrel/App.xaml.cs]: No, line 209 actually won't affect anything as the check has already happened.  Lines 209 and 227 should be removed.

File: src/Quarrel/App.xaml (hunk 7 above)
Comment (sharpninja) [reply src/Quarrel/App.xaml]: Windows themes don't work that way.  There are only three themes, Light, Dark and HighContrast.  We need to move to WinUI to get acrylic in Mobile and restore to the previous theme scheme.

File: src/Quarrel.ViewModels/Services/Gateway/GatewayService.cs (hunk 8 above)
Comment (sharpninja) [reply src/Quarrel.ViewModels/Services/Gateway/GatewayService.cs]: Preference, nothing else.

File: src/Quarrel.ViewModels/MainViewModel.cs (hunk 9 above)
Comment (sharpninja) [reply src/Quarrel.ViewModels/MainViewModel.cs]: This line at 310 of the previous version was actually redundant:

```csharp
bChannel.Typers[m.TypingStart.UserId].Change(0, 8 * 1000);
```
Look at the scope of the whole lambda

```csharp
                {
                    if (GuildsService.CurrentChannels.TryGetValue(m.TypingStart.ChannelId, out var bChannel))
                    {
                        if (bChannel.Typers.TryRemove(m.TypingStart.UserId, out var oldTimer))
                        {
                            oldTimer.Dispose();
                        }
                            
                        Timer timer = new Timer((s) =>
                        {
                            if (bChannel.Typers.TryRemove(m.TypingStart.UserId, out var oldUser))
                            {
                                oldUser.Dispose();
                            }

                            DispatcherHelper.CheckBeginInvokeOnUi(() =>
                            {
                                bChannel.RaisePropertyChanged(nameof(bChannel.IsTyping));
                                bChannel.RaisePropertyChanged(nameof(bChannel.TypingText));
                            });
                        }, null, 0, 8 * 1000);

                        bChannel.Typers.TryAdd(m.TypingStart.UserId, timer);
                        
                        DispatcherHelper.CheckBeginInvokeOnUi(() =>
                        {
                            bChannel.RaisePropertyChanged(nameof(bChannel.IsTyping));
                            bChannel.RaisePropertyChanged(nameof(bChannel.TypingText));
                        });
                    }
                });
```
Reusing the old timer would have the wrong UserID in the closure surrounding the lambda in the timer's event handler, so you need to always git rid of the old timer and make a new one.

Comment (sharpninja) [reply src/Quarrel.ViewModels/MainViewModel.cs]: The timer had to be recreated to ensure the closure around the tick event has the correct UserID when it fires.

**Comment by matthew4850** [comment]
Merged in #90 due to a few changes I made.
